proposals/reference-types/Overview.md

```diff
@@ -0,0 +1,204 @@
+  - In the binary format, space for the index is already reserved.
+  - For backwards compatibility, the index may be omitted in the text format, in which case it defaults to 0.
+
+
+Table extensions:
+
+* A module may define, import, and export multiple tables.
+  - As usual, the imports come first in the index space.
+  - This is already representable in the binary format.
+
+* Element segments take a table index as immediate that identifies the table they apply to.
+  - In the binary format, space for the index is already reserved.
+  - For backwards compatibility, the index may be omitted in the text format, in which case it defaults to 0.
+
+
+API extensions:
+
+* Any JS object (non-primitive value) or `null` can be passed as `anyref` to a Wasm function, stored in a global, or in a table.
+
+* Any JS function object or `null` can be passed as `anyfunc` to a Wasm function, stored in a global, or in a table.
+
+* Only `null` can be passed as a `nullref` to a Wasm function, stored in a global, or in a table.
+
+
+## Possible Future Extensions
+
+
+### Typed function references
+
+Motivation:
+
+* Allow function pointers to be expressed directly without going through table and dynamic type check.
+* Enable functions to be passed to other modules easily.
+
+Additions:
+
+* Add `(ref $t)` as a reference type
+  - `reftype ::= ... | ref <typeidx>`
+* Add `(ref.func $f)` and `(call_ref)` instructions
+  - `ref.func $f : [] -> (ref $t)  iff $f : $t`
+  - `call_ref : [ts1 (ref $t)] -> [ts2]` iff `$t = [ts1] -> [ts2]`
+* Introduce subtyping `ref <functype> < anyfunc`
+* Subtying between concrete and universal reference types
+  - `ref $t < anyref`
+  - `ref <functype> < anyfunc`
+
+* Typed function references cannot be null!
+
+Question:
+
+* General function have no reasonable default, do we need scoped variables like `let`?
+* Should there be a down cast instruction?
+* Should there be depth subtyping for function types?
+
+
+### Type Import/Export
+
+Motivation:
+
+* Allow the host (or Wasm modules) to distinguish different reference types.
+
+Additions:
+
+* Add `(type)` external type, enables types to be imported and exported
+  - `externtype ::= ... | type`
+  - `(ref $t)` can now denote an abstract type or a function reference
+  - imported types have index starting from 0.
+  - reserve byte in binary format to allow refinements later
+
+* Add abstract type definitions in type section
+  - `deftype ::= <functype> | new`
+  - creates unique abstract type
+
+* Add `WebAssembly.Type` class to JS API
+  - constructor `new WebAssembly.Type(name)` creates unique abstract type
+
+* Subtyping `ref <abstype>` < `anyref`
+
+Questions:
+
+* Do we need to impose constraints on the order of imports, to stratify section dependencies?
+
+* Do we need a nullable `(ref opt $t)` type to allow use with locals etc.?
```

Comment:
I definitely think we'll need nullable references for all the things (tables, import signatures, locals).  At that point, I wonder whether it's valuable (i.e., a measurable perf improvement when amortized by the overall call-to-JS/DOM) to have non-nullable types.  Maybe we could consider non-null types as a later backwards-compatible optimization?

Reply:
Hm, I would expect that in almost all cases other than certain initialisation patterns you don't want null to pollute your reference types, and you want to avoid the cost of an implicit null check for every single access (which may not be free on all architectures).

Reply:
I was mostly thinking about host types, but I see your point when we get to `ref` of GC types.  For GC types, it seems like there will be large classes of languages which both do and don't have ubiquitous nullable references.  Could we have "nullable" be either a type constructor (`(nullable (ref $T))`)?

Reply:
Possibly, good idea. It's basically the union `nullref | ref $T`, so a type constructor might work as well.

Reply:
Added an item on this. Thinking more about it, it's a bit less clear than it would seem at first, see added discussion.